storage/allocator.go

```diff
@@ -81,6 +93,120 @@ func getUsedNodes(existing []proto.Replica) map[proto.NodeID]struct{} {
+func (a *allocator) findDeadReplicas(desc *proto.RangeDescriptor) []proto.Replica {
+	var deadReplicas []proto.Replica
+	// TODO(mrtracy): Pull in Bram's storePool changes to support this.
+	/*
+		for _, repl := range desc.Replicas {
+			if a.storePool.getStoreDetail(repl.StoreID).dead {
+				deadReplicas = append(deadReplicas, repl)
+			}
+		}
+		return deadReplicas
+	*/
+	return deadReplicas
+}
+
+// rangeAvailability computes some availability statistics about a range that
+// has dead replicas.
+func (a *allocator) computeRangeAvailability(desc *proto.RangeDescriptor, dead []proto.Replica) (bool, int) {
+	quorum := (len(desc.Replicas) / 2) + 1
+	liveReplicas := len(desc.Replicas) - len(dead)
+	if liveReplicas < quorum {
+		return false, 0
+	}
+	return true, quorum - liveReplicas
+}
+
+// requiresRepair determines if the supplied range, as governed by the supplied
+// zone configuration, requires any sort of repair action. It returns
+// information about whether or not the range *can* be repaired, if the range
+// *should* be repaired, and the relative priority of the required repair (if
+// one is required).
+func (a *allocator) requiresRepair(zone config.ZoneConfig, desc *proto.RangeDescriptor) (
+	canRepair bool, needsRepair bool, priority float64) {
+	// Search for dead replicas first.
+	deadReplicas := a.findDeadReplicas(desc)
+	if len(deadReplicas) > 0 {
+		isAvailable, priority := a.computeRangeAvailability(desc, deadReplicas)
+		// Return immediately if range has so many dead replicas that it is
+		// unavailable.
+		if !isAvailable {
+			return false, true, 0
+		}
+
+		// Removal of dead replicas is the highest priority.
+		return true, true, float64(300 + priority)
+	}
+
+	// Return true if the current replica set doesn't match with zone config.
+	// TODO(mrtracy): Handle non-homogenous and mismatched attribute sets.
+	need := len(zone.ReplicaAttrs)
+	have := len(desc.Replicas)
+	if need == have {
+		return true, false, 0
+	}
+
+	neededQuorum := (need / 2) + 1
+	if have < need {
+		// Adding replicas is always a higher priority than removing replicas.
+		return true, true, float64(200 + (neededQuorum - have))
+	}
+
+	// Have > need. Ranges with an even number of replicas get extra priority
+	// because they have a more fragile quorum.
+	return true, true, float64(100 - (have % 2))
+}
+
+// computeRepair determines the exact operation needed to repair the supplied
+// range, as governed by the supplied zone configuration. It returns the
+// required action that should be taken and a replica on which the action should
+// be performed.
+func (a *allocator) computeRepair(zone config.ZoneConfig, desc *proto.RangeDescriptor) (
+	allocatorAction, proto.Replica, error) {
+	// Return remove dead node if any dead node.
+	deadReplicas := a.findDeadReplicas(desc)
+	if len(deadReplicas) > 0 {
+		isAvailable, _ := a.computeRangeAvailability(desc, deadReplicas)
+		if !isAvailable {
+			return aaUnavailable, proto.Replica{}, nil
```

Comment:
I think of it as two phases: the allocator computes a replication plan, and the queue executes it. Whether the range is available or not, removing dead nodes is part of the plan. The decision to not start a ChangeReplicas operation that we do not expect to succeed is part of the execution. It seems simpler to me to make one decision about availability right before taking action, instead of adding extra return values to thread availability information through the allocator.